Add format flag to network and volume ls

vdemeester opened the pull request:
This adds a `--format` flag to the `docker network ls` and `docker volume ls` commands, that
allows to customize the format of the output 🐹.

This also re-organize the `api/client/formatter` package by splitting it into several files (`container.go`, `image.go`, …).

This needs :
- [x] More unit tests on the `api/client/formatter` package
- [x] Integration tests
- [x] Documentation updates

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

thaJeztah commented:
Looks like there's a failure;

```
15:15:50 ----------------------------------------------------------------------
15:15:50 FAIL: docker_cli_run_test.go:4292: DockerSuite.TestRunNamedVolumeNotRemoved
15:15:50 
15:15:50 docker_cli_run_test.go:4300:
15:15:50     c.Assert(strings.TrimSpace(out), checker.Equals, "test")
15:15:50 ... obtained string = "Template parsing error: template: :1:2: executing \"\" at <.ID>: ID is not a field of struct type *formatter.volumeContext"
15:15:50 ... expected string = "test"
15:15:50 
15:15:51 
15:15:51 ----------------------------------------------------------------------
15:15:51 FAIL: docker_cli_run_test.go:4309: DockerSuite.TestRunNamedVolumesFromNotRemoved
15:15:51 
15:15:51 docker_cli_run_test.go:4322:
15:15:51     c.Assert(strings.TrimSpace(out), checker.Equals, "test")
15:15:51 ... obtained string = "Template parsing error: template: :1:2: executing \"\" at <.ID>: ID is not a field of struct type *formatter.volumeContext"
15:15:51 ... expected string = "test"
15:15:51 
```

thaJeztah commented:
Oh! No raw fallback, perhaps? `ID` -> `Id`

vdemeester commented:
@thaJeztah the volume type I use (`types.Volume`) have no idea (and there is no idea in `docker volume ls`), I did miss that :stuck_out_tongue_closed_eyes:

thaJeztah commented:
Think it makes sense to have this here as well, so moved to code review.

Already played with it a bit, and all looks to work properly (with some limited testing)

vdemeester commented on api/client/formatter/network.go:
```diff
@@ -0,0 +1,120 @@
+	return c.n.Driver
+}
+
+// YES/NO vs true/false vs 1/0 ?
```
@thaJeztah I wasn't sure what to put here for boolean values. How do we usually print them ?

thaJeztah replied:
Hm, good question, should have to check in other format outputs

vdemeester commented:
Updated with integration tests :angel: 
/cc @runcom @cpuguy83 @LK4D4 for code reviews :angel:

dnephin commented on api/client/cli.go:
```diff
@@ -123,6 +123,18 @@ func (cli *DockerCli) ImagesFormat() string {
+// String contains columns and format specification, for example {{ID}}\t{{Name}}
+func (cli *DockerCli) NetworksFormat() string {
+	return cli.configFile.NetworksFormat
+}
```
Do we need to expose all of these directly? I was really surprised to see an `ImageFormat()` method on `DockerCli`.

You can access this already from `dockerCli.ConfigFile().NetworksFormat`

vdemeester replied:
Hum true :sweat_smile: I just mimic the current behavior but I could rework them all to use `dockerCli.ConfigFile().NetworksFormat` I think :wink:

vdemeester commented:
Rebased and updated 👼

dnephin commented:
LGTM

dnephin commented:
retest this please

thaJeztah commented:
LGTM

thaJeztah commented:
@vdemeester can you add the docs changes to this PR?

vdemeester commented:
@thaJeztah on it 👼

thaJeztah commented:
Thanks! docs look good; can you update the man pages as well?

vdemeester commented:
@thaJeztah damn I forgot to update those :sweat_smile:

vdemeester committed:
Refactor api/client/formatter files

- Create container.go and image.go for specific context code
- Keep common code in formatter.go and custom.go

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester committed:
Add network --format flag to ls

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester committed:
Add volume --format flag to ls

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester committed:
Refactor dockerCli *Format to ConfigFile call

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

thaJeztah commented:
LGTM!

vdemeester commented:
Author LGTM (for docs :angel:)
